mesa/time.py

```diff
@@ -198,3 +198,78 @@ def step(self) -> None:
             self.time += self.stage_time
 
         self.steps += 1
+
+
+class RandomActivationByType(RandomActivation):
+    """
+    A scheduler which activates each type of agent once per step, in random
+    order, with the order reshuffled every step.
+
+    This is equivalent to the NetLogo 'ask breed...' and is generally the
+    default behavior for an ABM.
+
+    Assumes that all agents have a step() method.
+
+    This implementation assumes that the type of an agent doesn't change
+    throughout the simulation.
+    """
+
+    def __init__(self, model: Model) -> None:
+        super().__init__(model)
+        self.agents_by_type = defaultdict(dict)
+
+    def add(self, agent: Agent) -> None:
+        """
+        Add an Agent object to the schedule
+
+        Args:
+            agent: An Agent to be added to the schedule.
+        """
+
+        self._agents[agent.unique_id] = agent
+        agent_class: Type[Agent] = type(agent)
+        self.agents_by_type[agent_class][agent.unique_id] = agent
+
+    def remove(self, agent: Agent) -> None:
+        """
+        Remove all instances of a given agent from the schedule.
+        """
+
+        del self._agents[agent.unique_id]
+
+        agent_class: Type[Agent] = type(agent)
+        del self.agents_by_type[agent_class][agent.unique_id]
+
+    def step(self, by_type: bool = True) -> None:
+        """
+        Executes the step of each agent type, one at a time, in random order.
+
+        Args:
+            by_type: If True, run all agents of a single type before running
+                      the next one.
+        """
+        if by_type:
+            for agent_class in self.agents_by_type:
+                self.step_type(agent_class)
+            self.steps += 1
+            self.time += 1
+        else:
+            super().step()
+
+    def step_type(self, type_class: Type[Agent]) -> None:
+        """
+        Shuffle order and run all agents of a given type.
```

Comment:
This is equivalent to "ask breed" in [NetLogo](http://ccl.northwestern.edu/netlogo/docs/dict/breed.html) right? In that case, maybe explicitly document that.

Reply:
It's already stated in the main docstring: https://github.com/projectmesa/mesa/pull/1162/files#diff-2ae534e35a540eca09bc3f43b6e3ffd91695f0b178f4824869c8975a0c0423a5R208.

Reply:
I think this is a better place to (also) mention it since it's a function not available in other schedulers.

Reply:
OK, I additionally mentioned here, so it is mentioned twice, just to be sure.